### PR TITLE
[Yul] String performance

### PR DESCRIPTION
**Pull request by chriseth**
Closes https://github.com/ethereum/solidity/issues/5294

### CODE REVIEW

**Comment by leonardoalt** [comment]
`libyul/YulString.h` missing

**Comment by chriseth** [comment]
Added.

**Comment by ekpyron** [comment]
``libyul/optimiser/VarDeclPropagator.cpp`` still needs some ``.str()``'s it seems. One could consider making ``YulString`` implicitly convertible to ``std::string``? Maybe better to keep things explicit, though...

**Comment by chriseth** [comment]
I had implicit convertibility first but it is so easy to make it non-performant again due to that.

File: test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul [hunk 1]
```diff
@@ -17,19 +17,19 @@
 //         let length_1 := mload(from)
 //         length := length_1
 //         mstore(to, length_1)
-//         let from_1 := add(from, 0x20)
-//         let to_1 := add(to, 0x20)
+//         let from_2 := add(from, 0x20)
+//         let to_3 := add(to, 0x20)
```
Comment (leonardoalt) [inline test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul]: Why do these indices start with 2 and 3?

Comment (leonardoalt) [reply test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul]: It looks like the SSA index is global for all variables (of a certain scope).

Comment (ekpyron) [reply test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul]: The ``NameDispenser`` now seems to have a global counter - but what's the reason for doing that? In any case the result is still valid...

Comment (chriseth) [reply test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul]: Yeah, I will split that out into a different PR. This is #5074

**Review by leonardoalt — DISMISSED** [review]

File: test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul (hunk 1 above)
Comment (leonardoalt) [reply test/libyul/yulOptimizerTests/ssaPlusCleanup/control_structures.yul]: I think it's fine in the same PR.

**Comment by chriseth** [comment]
Hey, I haven't split it out yet!

**Comment by chriseth** [comment]
Now I have :)

**Comment by chriseth** [comment]
Oh wow, I actually did already change it back, but I only forgot the tests....

**Pull request merged**